experimental/dds/tree2/src/core/tree/visitDelta.ts

```diff
@@ -361,368 +302,136 @@ function offsetDetachId(
-				for (const { oldContent, newContent } of replaces) {
-					if (oldContent !== undefined) {
-						visitNode(index, oldContent.fields, visitor, config);
-						if (newContent === undefined) {
-							// This a simple detach
-							const oldRoot = oldContent.destination;
-							const field = config.detachedFieldIndex.toFieldKey(brand(oldRoot));
-							visitor.detach({ start: index, end: index + 1 }, field);
-						} else {
-							// This really is a replace.
-							// Delay the detach until we can do it during the attach pass.
-							index += 1;
+			if (mark.detach !== undefined) {
+				if (mark.attach === undefined) {
+					// This a simple detach
+					for (let i = 0; i < mark.count; i += 1) {
+						const root = config.detachedFieldIndex.getOrCreateEntry(
+							offsetDetachId(mark.detach, i),
+						);
+						if (mark.fields !== undefined) {
+							config.attachPassRoots.set(root, mark.fields);
 						}
+						const field = config.detachedFieldIndex.toFieldKey(root);
+						visitor.detach({ start: index, end: index + 1 }, field);
 					}
+				} else {
+					// This really is a replace.
+					// Delay the detach until we can do it during the attach pass.
+					index += mark.count;
 				}
+			} else if (mark.attach === undefined) {
+				index += mark.count;
 			}
 		}
 	}
 }
 
 /**
  * Preforms the following:
- * - Collects all roots that need an attach pass
- * - Executes attaches (top-down)
- * - Executes replaces (top-down)
+ * - Executes attaches (top-down) applying nested changes on the attached nodes
+ * - Executes replaces (top-down) applying nested changes on the attached nodes
+ * - Collects detached roots (from replaces) that need an attach pass
  */
-function attachPass(delta: Delta.MarkList, visitor: DeltaVisitor, config: PassConfig): void {
-	let index = 0;
-	for (const mark of delta) {
-		if (typeof mark === "number") {
-			// Untouched nodes
-			index += mark;
-		} else {
-			catalogAttachPassRootChanges(mark, config);
-			if (mark.type === Delta.MarkType.Modify) {
-				if (mark.detachedNodeId === undefined) {
-					visitNode(index, mark.fields, visitor, config);
-					index += 1;
-				}
-			} else {
-				const replaces = asReplaces(mark, config);
-				for (const { oldContent, newContent } of replaces) {
-					if (newContent !== undefined) {
-						const newRoot = newContent.source;
-						const newContentField = config.detachedFieldIndex.toFieldKey(newRoot);
-						if (oldContent !== undefined) {
+function attachPass(delta: Delta.FieldChanges, visitor: DeltaVisitor, config: PassConfig): void {
+	if (delta.attached !== undefined) {
+		let index = 0;
+		for (const mark of delta.attached) {
+			if (mark.detach === undefined || mark.attach !== undefined) {
+				if (mark.attach !== undefined) {
+					for (let i = 0; i < mark.count; i += 1) {
+						const offsetAttachId = offsetDetachId(mark.attach, i);
+						const sourceRoot =
+							config.detachedFieldIndex.getOrCreateEntry(offsetAttachId);
```

Comment:
```suggestion
							config.detachedFieldIndex.getEntry(offsetAttachId);
```

The entry should always be defined here, right?